src/utils/units.hpp

```diff
@@ -0,0 +1,142 @@
+//========================================================================================
+// (C) (or copyright) 2024. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_UNITS_HPP_
+#define UTILS_UNITS_HPP_
+
+#include "artemis.hpp"
+
+#include <cmath>
+
+namespace ArtemisUtils {
+
+enum class UnitSystem { scalefree, cgs };
+
+class Units {
+ public:
+  // No default constructor
+  Units() = delete;
+  // Host-only constructor from parameter input file
+  Units(ParameterInput *pin, std::shared_ptr<StateDescriptor> pkg);
+
+  // Copy constructor must be marked with KOKKOS_FUNCTION
+  KOKKOS_FUNCTION
+  Units(const Units &other) {}
+
+  // Return unit system
+  KOKKOS_INLINE_FUNCTION
+  UnitSystem GetUnitSystem() const { return unit_system_; }
+
+  // Unit conversions
+  KOKKOS_INLINE_FUNCTION
+  Real GetLengthCodeToPhysical() const { return length_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetLengthPhysicalToCode() const { return 1. / length_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetTimeCodeToPhysical() const { return time_; }
```

Comment:
What do you think about these functions taking in the quantity and returning the converted quantity, as opposed to returning conversion factors and the caller applying them.

Reply:
I think this is a design choice that doesn't warrant a rewrite at this point. The one issue with the returning-converted-quantities approach is that you *need* functions for every kind of quantity you want to convert, unless you want to break the design if necessary by doing like
```c++
double l_unit = ConvertLengthCodeToCGS(1.);
double t_unit = ConvertTimeCodeToCGS(1.);
// Use some combination of powers of l_unit and t_unit to convert your custom quantity
```

Reply:
I agree my choice comes with needing functions for every kind of quantity. I think that comes with safety, though, since there is one implementation of a unit conversion, and not multiple implementations. If a developer needs a conversion function that doesn't exist yet, they can add it.

Reply:
As discussed today let's keep the current function signatures.